src/file_scanner/arrow_multi_file_info.cpp

```diff
@@ -0,0 +1,244 @@
+
+#include "ipc/stream_reader/ipc_file_stream_reader.hpp"
+
+#include "duckdb/common/bind_helpers.hpp"
+#include "file_scanner/arrow_file_scan.hpp"
+#include "ipc/stream_factory.hpp"
+
+namespace duckdb {
+namespace ext_nanoarrow {
+
+unique_ptr<BaseFileReaderOptions> ArrowMultiFileInfo::InitializeOptions(
+    ClientContext& context, optional_ptr<TableFunctionInfo> info) {
+  return make_uniq<ArrowFileReaderOptions>();
+}
+
+bool ArrowMultiFileInfo::ParseCopyOption(ClientContext& context, const string& key,
+                                         const vector<Value>& values,
+                                         BaseFileReaderOptions& options_p,
+                                         vector<string>& expected_names,
+                                         vector<LogicalType>& expected_types) {
+  return true;
+}
+
+bool ArrowMultiFileInfo::ParseOption(ClientContext& context, const string& key,
+                                     const Value& val, MultiFileOptions& file_options,
+                                     BaseFileReaderOptions& options) {
+  return true;
+}
+
+void ArrowMultiFileInfo::FinalizeCopyBind(ClientContext& context,
+                                          BaseFileReaderOptions& options_p,
+                                          const vector<string>& expected_names,
+                                          const vector<LogicalType>& expected_types) {}
+
+struct ArrowMultifileData : public TableFunctionData {
+  ArrowMultifileData() = default;
+
+  unique_ptr<ArrowFileScan> file_scan;
+};
+
+unique_ptr<TableFunctionData> ArrowMultiFileInfo::InitializeBindData(
+    MultiFileBindData& multi_file_data, unique_ptr<BaseFileReaderOptions> options_p) {
+  return make_uniq<ArrowMultifileData>();
+}
+
+void ArrowMultiFileInfo::BindReader(ClientContext& context,
+                                    vector<LogicalType>& return_types,
+                                    vector<string>& names, MultiFileBindData& bind_data) {
+  auto& arrow_multifile_data = bind_data.bind_data->Cast<ArrowMultifileData>();
+  auto& multi_file_list = *bind_data.file_list;
+  if (!bind_data.file_options.union_by_name) {
+    arrow_multifile_data.file_scan =
+        make_uniq<ArrowFileScan>(context, multi_file_list.GetFirstFile());
+    return_types = arrow_multifile_data.file_scan->GetTypes();
+    names = arrow_multifile_data.file_scan->GetNames();
+    bind_data.multi_file_reader->BindOptions(bind_data.file_options, multi_file_list,
+                                             return_types, names, bind_data.reader_bind);
+  } else {
+    ArrowFileReaderOptions options;
+    bind_data.reader_bind =
+        bind_data.multi_file_reader->BindUnionReader<ArrowMultiFileInfo>(
+            context, return_types, names, multi_file_list, bind_data, options,
+            bind_data.file_options);
+  }
+  D_ASSERT(names.size() == return_types.size());
+}
+
+void ArrowMultiFileInfo::FinalizeBindData(MultiFileBindData& multi_file_data) {}
```

Comment:
To continue on the previous comment, about using `MultiFileReader::BindReader`, when not using `union_by_name`, it will initialize the `MultiFileBindData` with the reader created to bind:
```c++
result.Initialize(std::move(reader));
```

Then, in FinalizeBindData, the parquet implementation does this:
```c++
	auto &bind_data = multi_file_data.bind_data->Cast<ParquetReadBindData>();
	if (multi_file_data.initial_reader) {
		auto &initial_reader = multi_file_data.initial_reader->Cast<ParquetReader>();
		bind_data.initial_file_cardinality = initial_reader.NumRows();
		bind_data.initial_file_row_groups = initial_reader.NumRowGroups();
		bind_data.parquet_options = initial_reader.parquet_options;
	}
```
To reuse this "initial_reader" instead of creating a new one